share/harvest/base.py

```diff
@@ -1,196 +1,250 @@
+            raise ValueError('To ensure that data is optimally deduplicated, harvests may not occur while using a pretty serializer.')
 
-    def _validate_dates(self, start_date, end_date):
-        assert not (bool(start_date) ^ bool(end_date)), 'Must specify both a start and end date or neither'
-        assert isinstance(start_date, (datetime.timedelta, datetime.datetime, pendulum.Pendulum)) and isinstance(end_date, (datetime.timedelta, datetime.datetime, pendulum.Pendulum)), 'start_date and end_date must be either datetimes or timedeltas'
-        assert not (isinstance(start_date, datetime.timedelta) and isinstance(end_date, datetime.timedelta)), 'Only one of start_date and end_date may be a timedelta'
+        if (self.config.disabled or self.config.source.is_deleted) and not (force or ignore_disabled):
+            raise HarvesterDisabledError('Harvester {!r} is disabled. Either enable it, run with force=True, or ignore_disabled=True.'.format(self.config))
 
-        if isinstance(start_date, datetime.datetime):
-            start_date = pendulum.instance(start_date)
+        with self.config.acquire_lock(required=not force):
+            logger.info('Harvesting %s - %s from %r', start, end, self.config)
+            yield from RawDatum.objects.store_chunk(self.config, self.fetch_date_range(start, end, **kwargs), limit=limit)
 
-        if isinstance(end_date, datetime.datetime):
-            end_date = pendulum.instance(end_date)
+    def harvest_from_log(self, harvest_log, **kwargs):
+        """Harvest data as specified by the given harvest_log.
 
-        if isinstance(start_date, datetime.timedelta):
-            start_date = pendulum.instance(end_date + start_date)
+        Args:
+            harvest_log (HarvestLog): The HarvestLog that describes the parameters of this harvest
+            limit (int, optional): The maximum number of unique data to harvest. Defaults to None.
+            **kwargs: Forwared to harvest_date_range.
 
-        if isinstance(end_date, datetime.timedelta):
-            end_date = pendulum.instance(start_date + end_date)
+        Yields:
+            RawDatum
 
-        og_start, og_end = start_date, end_date
-        start_date, end_date = self.shift_range(start_date, end_date)
-        assert isinstance(start_date, pendulum.Pendulum) and isinstance(end_date, pendulum.Pendulum), 'transpose_time_window must return a tuple of 2 datetimes'
+        """
+        error = None
+        datum_ids = []
+        logger.info('Harvesting %r', harvest_log)
+
+        with harvest_log.handle(self.VERSION):
+            try:
+                for datum in self.harvest_date_range(harvest_log.start_date, harvest_log.end_date, **kwargs):
+                    datum_ids.append(datum.id)
+                    yield datum
+            except Exception as e:
+                error = e
+                raise error
+            finally:
+                try:
+                    harvest_log.raw_data.add(*datum_ids)
```

Comment:
It seems like we should associate raws with the log right when they're created, in `store_chunk`, rather than waiting until the entire harvest is done, transform tasks are already created, and everything.